arangod/Agency/Agent.cpp

```diff
@@ -1155,15 +1182,49 @@ void Agent::run() {
       _agentNeedsWakeup = false;
     }
 
+    if (leading() && getPrepareLeadership() == 1) {
+      // If we are officially leading but the _preparing flag is set, we
+      // are in the process of preparing for leadership. This flag is
+      // set when the Constituent celebrates an election victory. Here,
+      // in the main thread, we do the actual preparations:
+
+      if (!prepareLead()) {
+        _constituent.follow(0);  // do not change term
+      } else {
+        // we need to start work as leader
+        lead();
+      }
+
+      donePrepareLeadership();   // we are ready to roll, except that we
+                                 // have to wait for the _commitIndex to
+                                 // reach the end of our log
+    }
+
     // Leader working only
     if (leading()) {
```

Comment:
Similar race condition, reversed logic:  leading() could have just been set and getPrepareLeadership() was zero in previous "if", but is now one.  Suggest an "&& 0==getPrepareLeadership()" test added to "if" condition.  Also, add an else clause that sets _agentNeedsWakup so that loop repeats and performs these two steps in order.

Reply:
Again, yes, but I think it is a bit more subtle here. _preparing is set to 1 when we become leader. Then the preparation code runs which sets it to 2 (donePrepareLeadership), we then have to run the main work which will increase `_commitIndex` and only when that reaches the end of the log may we set _preparing to 0, finally terminating the preparation. Therefore, the additional condition needs to be `if (1 == getPrepareLeadership()) { continue; }` which I added.